File: dynamic_programming/UniquePathsWithObstacles.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        width = len(obstacleGrid[0])
        list = [0] * width
        list[0] = 1
        for row in obstacleGrid:
            for i in range (width):
                if row[i] == 1:
                    list[i] = 0
                elif i>0:
                    list[i] += list[i-1]
        return list[width-1]

File: dynamic_programming/UniquePathsWithObstacles.cs
public class Solution {
    public int UniquePathsWithObstacles(int[][] obstacleGrid) {
        // define the width
        int width = obstacleGrid[0].Length;
        var list = Enumerable.Repeat(0, width).ToArray();
        // robot position
        list[0] = 1;
        foreach (int[] row in obstacleGrid){
            for(int i=0; i<width; i++){
                // if the rock is at index i
                if(row[i] == 1)
                    list[i] = 0;
                // make sure i is not 0
                else if(i>0)
                    list[i] += list[i-1];
            }
        }
        return list.Last();
    }
}